src/ledger-bridge-keyring.ts

```diff
@@ -0,0 +1,908 @@
+}
+
+interface IFrameMessage {
+  action: IFrameMessageAction;
+  params: Record<string, unknown>;
+  target: string;
+  messageId: number;
+}
+
+interface IFrameMessageGetAddressResponsePayload {
+  publicKey: string;
+  address: string;
+  chainCode: string;
+  error?: string;
+}
+
+interface IFrameMessageSignResponsePayload {
+  v: string;
+  r: string;
+  s: string;
+  error?: unknown;
+}
+
+interface IFrameMessageConnectionResponsePayload {
+  connected: boolean;
+  error?: unknown;
+}
+
+type IFrameMessageResponsePayload =
+  | IFrameMessageGetAddressResponsePayload
+  | IFrameMessageSignResponsePayload
+  | IFrameMessageConnectionResponsePayload;
+
+interface IFrameMessageResponse {
+  success: boolean;
+  action: IFrameMessageAction;
+  messageId: number;
+  payload: IFrameMessageResponsePayload;
+  error?: unknown;
+}
+
+export interface AccountDetails {
+  index?: number;
+  bip44?: boolean;
+  hdPath?: string;
+}
+
+export interface LedgerBridgeKeyringOptions {
+  hdPath?: string;
+  accounts?: string[];
+  accountDetails?: Record<string, AccountDetails>;
+  accountIndexes?: Record<string, number>;
+  bridgeUrl?: string;
+  implementFullBIP44?: boolean;
+}
+
+/**
+ * Check if the given transaction is made with ethereumjs-tx or @ethereumjs/tx
+ *
+ * Transactions built with older versions of ethereumjs-tx have a
+ * getChainId method that newer versions do not.
+ * Older versions are mutable
+ * while newer versions default to being immutable.
+ * Expected shape and type
+ * of data for v, r and s differ (Buffer (old) vs BN (new)).
+ *
+ * @param tx
+ * @returns Returns `true` if tx is an old-style ethereumjs-tx transaction.
+ */
+function isOldStyleEthereumjsTx(
+  tx: TypedTransaction | OldEthJsTransaction,
+): tx is OldEthJsTransaction {
+  return typeof (tx as OldEthJsTransaction).getChainId === 'function';
+}
+
+function isSignMessageResponse(
+  payload: IFrameMessageResponsePayload,
+): payload is IFrameMessageSignResponsePayload {
+  return typeof (payload as IFrameMessageSignResponsePayload).v === 'string';
+}
+
+function isGetAddressMessageResponse(
+  payload: IFrameMessageResponsePayload,
+): payload is IFrameMessageGetAddressResponsePayload {
+  return (
+    typeof (payload as IFrameMessageGetAddressResponsePayload).publicKey ===
+    'string'
+  );
+}
+
+function isConnectionChangedResponse(
+  payload: IFrameMessageResponsePayload,
+): payload is IFrameMessageConnectionResponsePayload {
+  return (
+    typeof (payload as IFrameMessageConnectionResponsePayload).connected ===
+    'boolean'
+  );
+}
+
+export class LedgerBridgeKeyring extends EventEmitter {
+  static type: string = keyringType;
+
+  readonly type: string = keyringType;
+
+  page = 0;
+
+  perPage = 5;
+
+  unlockedAccount = 0;
+
+  accounts: readonly string[] = [];
+
+  accountDetails: Record<string, AccountDetails> = {};
+
+  hdk = new HDKey();
+
+  hdPath = hdPathString;
+
+  paths: Record<string, number> = {};
```

Comment:
These `Record`'s for `this.paths` and `this.accountDetails` could be made readonly via `Readonly<Record<...>>`, the only thing it has very slight chance of breakage depending on how the app is handled. Maybe best to allow these objects to be mutable